### PR TITLE
Fix: Improve allowedHostnames .env configuration and fail-fast in production

### LINKED ISSUE #4480: [Bug]: .env app.allowedHostNames does not work as intended
### Bug Description?

In the app\Config\App.php class the getValidHost() function checks $this->allowedHostNames to see if it's configured then logs a message saying to override it in .env but .env cannot override or add to an array.  This has been reported here (https://github.com/codeigniter4/CodeIgniter4/issues/7311).  To resolve the issue we need to implement something like this (https://github.com/codeigniter4/CodeIgniter4/pull/5526/changes) to override the contents of $this->allowedHostNames before checking it to see if it's populated.
```
    private function getValidHost(): string
    {
        $httpHost = $_SERVER['HTTP_HOST'] ?? 'localhost';

        if (empty($this->allowedHostnames)) {
            log_message('warning',
                'Security: allowedHostnames is not configured. ' .
                'Host header injection protection is disabled. ' .
                'Please set app.allowedHostnames in your .env file. ' .
                'Received Host: ' . $httpHost
            );
            return 'localhost';
        }

        if (in_array($httpHost, $this->allowedHostnames, true)) {
            return $httpHost;
        }

        log_message('warning',
            'Security: Rejected HTTP_HOST "' . $httpHost . '" - not in allowedHostnames whitelist. ' .
            'Using fallback: ' . $this->allowedHostnames[0]
        );

        return $this->allowedHostnames[0];
    }
```
Something like this:
```
        $envAllowedHostNames = getenv('app.allowedHostNames');
        if ($envAllowedHostNames)
        {
            $this->allowedHostNames = array_map('trim', explode(',', $allowedHostNames));
        }
```
Then in the .env we have something like `app.allowedHostNames = 'foo.com, localhost';` would resolve this issue.

### Steps to Reproduce?

1. populate app.allowedHostNames.0, et. al. in .env
2. Run the application
3. check the logs to see the warning.
4. Smash your palm to your face.

### Expected Behavior?

.env app.allowedHostNames to override the value of app\Config\App.php->allowedHostNames[]

### OpensourcePOS Version

development (unreleased)

### Php version

Php 8.4

### What browsers are you seeing the problem on?

Firefox

### Server Operating System and version

Windows 11 Professional

### Database Management System and version

MySQL 8.4.3

### Web Server and version

Apache 2.4.62

### System Information Report (optional)

System Information Report

### Unmodified copy of OpensourcePOS

- [x] I agree this copy has not been modified

### PR DESCRIPTION
**Pull request by jekkos**
## Summary

- **Fixes #4480**: `.env` `app.allowedHostnames` now works correctly via comma-separated values
- **Security Improvement**: Production environments now **fail explicitly** if `allowedHostnames` is not configured
- **Documentation**: Clear setup instructions in `INSTALL.md` with error messages

## Problem

The original implementation had a critical flaw:
1. **CodeIgniter 4 cannot override array properties from `.env`** ([codeigniter4/CodeIgniter4#7311](https://github.com/codeigniter4/CodeIgniter4/issues/7311))
2. Setting `app.allowedHostnames.0`, `app.allowedHostnames.1` did NOT populate the array
3. The `allowedHostnames` array was ALWAYS empty, so the app ALWAYS fell back to `localhost`
4. **Silent failure** in production - URLs generated as `http://localhost/` instead of actual domain
5. Host header injection protection (GHSA-jchf-7hr6-h4f3) was effectively disabled

This was reported in #4480 where users couldn't configure `allowedHostnames` properly.

### 1. Support Comma-Separated `.env` Values

```bash
# NEW FORMAT (works!)
app.allowedHostnames = 'example.com,www.example.com,demo.opensourcepos.org'

# OLD FORMAT (did NOT work)
app.allowedHostnames.0 = 'example.com'
app.allowedHostnames.1 = 'www.example.com'
```

### 2. Fail-Fast in Production

```php
// Production environment: Explicit RuntimeException
if (empty($this->allowedHostnames) && $environment === 'production') {
    throw new RuntimeException(
        'Security: allowedHostnames is not configured. ' .
        'Set app.allowedHostnames in your .env file.'
    );
}

// Development: Allow localhost fallback with ERROR logging
if (empty($this->allowedHostnames)) {
    log_message('error', '...Using localhost fallback (development only).');
    return 'localhost';
}
```

### 3. Clear Error Messages

When `allowedHostnames` is not configured in production:
```
RuntimeException: Security: allowedHostnames is not configured.
Host header injection protection is disabled.
Set app.allowedHostnames in your .env file.
Example: app.allowedHostnames = "example.com,www.example.com"
```

## Changes

| File | Change |
|------|--------|
| `app/Config/App.php` | Parse comma-separated `.env` values, fail in production, enhanced error logging |
| `.env.example` | Updated format documentation (comma-separated) |
| `INSTALL.md` | Prominent security section with clear instructions |
| `tests/Config/AppTest.php` | Comprehensive tests for new behavior |

## Breaking Changes

⚠️ **Deployments without `allowedHostnames` configured will now fail in production.**

This is **intentional** - silent failures are a security risk. Users must explicitly configure allowed hostnames.

**Migration Path:**
```bash
# Add to .env (comma-separated)
app.allowedHostnames = 'yourdomain.com,www.yourdomain.com'
```

## Testing

- [x] Comma-separated `.env` parsing works correctly
- [x] Production throws exception when not configured
- [x] Development allows localhost fallback with ERROR logging
- [x] Invalid hosts fall back to first configured hostname
- [x] Single hostname values work correctly
- [x] Whitespace trimming on comma-separated values

## Related

- **Issue**: #4480
- **Security Advisory**: GHSA-jchf-7hr6-h4f3
- **Upstream Issue**: [codeigniter4/CodeIgniter4#7311](https://github.com/codeigniter4/CodeIgniter4/issues/7311)

## Industry Best Practices

This implementation follows patterns from:
- **Django**: `ALLOWED_HOSTS` raises `DisallowedHost` in production
- **Rails**: `HostAuthorization` middleware returns 403 Forbidden
- **ASP.NET Core**: Host Filtering requires explicit configuration

All major frameworks agree: **fail explicitly in production, don't silently fall back.**

## Summary by CodeRabbit

* **Documentation**
  * Clarified that allowed hostnames are REQUIRED for production, updated examples to a single comma-separated format, and set an explicit empty default example.

* **Bug Fixes / Security**
  * Production now fails to start if allowed hostnames are not configured; development logs an error and falls back to localhost.

* **Tests**
  * Added tests for comma-separated parsing, trimming, single/empty values, filtering empty entries, and production vs. development behavior.

### CODE REVIEW

File: app/Config/App.php [hunk 1]
```diff
@@ -291,6 +291,15 @@ class App extends BaseConfig
     public function __construct()
     {
         parent::__construct();
+        
+        // Support comma-separated .env configuration for allowedHostnames
+        // Workaround for CodeIgniter 4 limitation: arrays cannot be set from .env
```
Comment (objecttothis) [inline app/Config/App.php]:
```suggestion
        // Solution for CodeIgniter 4 limitation: arrays cannot be set from .env
```
The word Workaround suggests a temporary solution that adds technical debt.

File: app/Config/App.php [hunk 2]
```diff
@@ -291,6 +291,15 @@ class App extends BaseConfig
...
+        // See: https://github.com/codeigniter4/CodeIgniter4/issues/7311
+        $envAllowedHostnames = getenv('app.allowedHostnames');
+        if ($envAllowedHostnames !== false && !empty($envAllowedHostnames)) {
+            $this->allowedHostnames = array_map('trim', explode(',', $envAllowedHostnames));
```
[the remark this answers is not in the document]
Comment (objecttothis) [reply app/Config/App.php]: This is a good suggestion. If we implement it we should also add a unit test to test for it.

**Review by objecttothis — CHANGES_REQUESTED** [review]

**Comment by jekkos** [comment]
## ✅ Review Feedback Addressed

Thank you for the detailed code review! I've incorporated all feedback in commit 671490d88:

### **CodeRabbit Suggestions**

**1. Filter empty hostnames from comma-separated values** ✅
- Changed from simple  to 
- Now properly handles trailing commas, leading commas, and whitespace-only entries
- Added regression tests for all edge cases

**2. Add language identifier to code fence** ✅  
- Changed ``` to ```text in INSTALL.md
- Resolves markdownlint MD040 warning

**3. Reset all $_SERVER keys in tearDown** ✅
- Added unset for SCRIPT_NAME and HTTPS
- Prevents test order dependency issues

### **objecttothis Feedback**

**1. Workaround → Solution** ✅
- Updated comment to reflect this is the permanent solution, not temporary

**2. Add unit tests for empty filtering** ✅
- Added comprehensive test: `testEnvAllowedHostnamesFiltersEmptyEntries`
- Tests trailing comma, leading comma, whitespace-only, and all-whitespace values

### **Tests Added**
```php
public function testEnvAllowedHostnamesFiltersEmptyEntries(): void
{
    // Tests: trailing comma, leading comma, whitespace entries, all-whitespace
}
```

All CI checks passing ✅

**Review by objecttothis — APPROVED** [review]
